### PR TITLE
Add python 3.12 support

### PR DESCRIPTION
**Pull request by sjdilkes**
# Checklist

- [ ] I have performed a self-review of my code.
- [ ] I have commented hard-to-understand parts of my code.
- [ ] I have made corresponding changes to the public API documentation.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have updated the changelog with any user-facing changes.

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -29,7 +29,10 @@ jobs:
           - { python: "3.11", os: "windows-2022", session: "tests" }
           - { python: "3.11", os: "ubuntu-22.04", session: "tests" }
           - { python: "3.11", os: "macos-12", session: "tests" }
-          - { python: "3.10", os: "ubuntu-22.04", session: "docs-build" }
+          - { python: "3.12", os: "windows-2022", session: "tests" }
```
Comment (cqc-melf) [inline .github/workflows/tests.yml:32]: I think it would be good to update the precommit and mypy checks to run with 3.11 now?

Comment (sjdilkes) [reply .github/workflows/tests.yml:32]: Just changed!

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -22,8 +22,8 @@ jobs:
       fail-fast: false
       matrix:
         include:
-          - { python: "3.10", os: "ubuntu-22.04", session: "pre-commit" }
-          - { python: "3.10", os: "macos-12", session: "mypy" }
+          - { python: "3.11", os: "ubuntu-22.04", session: "pre-commit" }
```
Comment (cqc-melf) [inline .github/workflows/tests.yml:25]:
```suggestion
          - { python: "3.11", os: "ubuntu-22.04", session: "pre-commit" }
          - { python: "3.10", os: "ubuntu-22.04", session: "tests" }
          - { python: "3.10", os: "macos-12", session: "tests" }

Comment (cqc-melf) [reply .github/workflows/tests.yml:25]: And adding normal tests runs for 3.10 with the corresponding os

**Review by cqc-melf — APPROVED** [review]

**Pull request merged**